.github/workflows/main.yml

```diff
@@ -86,24 +263,80 @@ jobs:
           MYSQL_PASSWORD: ibis
         options: --health-cmd="mysqladmin ping" --health-interval 10s --health-timeout 5s --health-retries 3
     steps:
-    - name: checkout
-      uses: actions/checkout@v1
-
-    - name: set up environment
-      run: ./ci/setup_env.sh "${{ matrix.python_version }}" "$BACKENDS"
-
-    - name: run tests
-      run: PYTEST_BACKENDS=$BACKENDS ./ci/run_tests.sh -m "not udf"
-
-    - name: pytest errors
-      run: ./ci/pytest_errors.sh
-      if: failure()
-
-  Tests_impala_clickhouse:
-    name: Tests Impala / Clickhouse
+      - name: checkout
+        uses: actions/checkout@v2
+
+      - name: download env file
+        uses: actions/download-artifact@v2
+        with:
+          name: conda-env-file
+          path: tmp
+
+      - uses: conda-incubator/setup-miniconda@v2
+        with:
+          mamba-version: "*"
+          miniforge-variant: Mambaforge
+          miniforge-version: latest
+          channel-priority: strict
+          activate-environment: ibis
+          python-version: ${{ matrix.python-version }}
+          environment-file: tmp/environment.yaml
+          condarc-file: ci/condarc
+
+      - name: install pyspark
+        run: |
+          conda install -n ibis \
+            pyspark=${{ matrix.pyspark.version }} \
+            pyarrow=${{ matrix.pyarrow }} \
+            openjdk="${{ matrix.pyspark.jdk }}"
+
+      - name: install ibis
+        run: pip install .
+
+      - name: download backend data
+        run: python ci/datamgr.py download
+
+      - name: set backends
+        id: set_backends
+        run: echo '::set-output name=backends::["mysql", "postgres", "pyspark"]'
+
+      - name: install backend data
+        run: |
+          set -euo pipefail
+
+          for backend in ${{ join(fromJSON(steps.set_backends.outputs.backends), ' ') }}; do
+            python ci/datamgr.py "$backend"
+          done
+
+      - name: run tests
+        env:
+          PYTEST_BACKENDS: ${{ join(fromJSON(steps.set_backends.outputs.backends), ' ') }}
+          ARROW_PRE_0_15_IPC_FORMAT: ${{ matrix.pyspark.env.ARROW_PRE_0_15_IPC_FORMAT }}
+        run: |
+          pytest \
+            --numprocesses auto \
+            ibis/backends/{${{ join(fromJSON(steps.set_backends.outputs.backends), ',') }}} \
+            ibis/backends/tests \
+            --durations=25 \
+            -m "(postgres and not udf) or mysql or pyspark" \
+            --quiet \
+            -ra \
+            --junitxml=junit.xml \
+            --cov=ibis \
+            --cov-report=xml:coverage.xml | tee pytest.log
+
+      - name: pytest errors
+        run: ./ci/pytest_errors.sh
+        if: failure()
+
+  test_impala_clickhouse:
```

Comment:
I'd prefer having separate jobs for impala and clickhouse, so we can directly see which backend is failing from github checks.

Reply:
Actually I'd prefer this approach for the simple backends as well by parametrizing the job matrix per backend. GHA provides enough parallelization so we don't need to worry about redundant install steps, which should be pretty click using poetry.

Reply:
done